clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -310,6 +310,7 @@ public void reenqueue(ProducerBatch batch, long now) {
         Deque<ProducerBatch> deque = getOrCreateDeque(batch.topicPartition);
         synchronized (deque) {
             deque.addFirst(batch);
+            maybeEnsureQueueIsOrdered(deque, batch);
```

Comment:
Mentioned previously, but I think it would be more intuitive to insert in order rather than fixing up the collection after the insertion.

Reply:
Would like to hear your thoughts on my previous comment as well:
>Another thing I was wondering is whether a deque is still the right data structure given that we now require ordering. Would a priority queue give us everything we need?